### PR TITLE
Add 42 School provider

### PR DESCRIPTION
**Pull request by awbx**
refer to #108

### CODE REVIEW

File: src/providers/42.ts [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+			refreshToken: result.refresh_token,
+			expiresIn: result.expires_in,
+			scope: result.scope,
+			tokenType: result.token_type
+		};
+		return tokens;
+	}
+}
+
+export type Scope =
```
Comment (pilcrowonpaper) [inline src/providers/42.ts]: Can you remove this to be consistent with other providers?

File: src/providers/42.ts [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+	| "elearning" // manage media related stuff
+	| "tig" // manage community services
+	| "forum"; // manage topics and messages
+
+export interface FortyTwoTokens {
+	accessToken: string;
+	refreshToken?: string | null;
+	expiresIn?: number;
+	tokenType?: string;
+	scope?: string;
```
Comment (pilcrowonpaper) [inline src/providers/42.ts]: I don't think `scope` required here

File: src/providers/42.ts [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+	| "public" // access the user public data
+	| "projects" // manage teams, slots and all projects related stuff
+	| "profile" // manage user data
+	| "elearning" // manage media related stuff
+	| "tig" // manage community services
+	| "forum"; // manage topics and messages
+
+export interface FortyTwoTokens {
+	accessToken: string;
+	refreshToken?: string | null;
```
Comment (pilcrowonpaper) [inline src/providers/42.ts]: Can you replace `a?: A` with `a: A | null`?

File: src/providers/42.ts [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+	constructor(
+		clientId: string,
+		clientSecret: string,
+		options?: {
+			redirectURI?: string;
+		}
+	) {
+		const baseUrl = "https://api.intra.42.fr";
+
+		const authorizeEndpoint = baseUrl + "/oauth/authorize";
```
Comment (pilcrowonpaper) [inline src/providers/42.ts]: Super minor nit-pick but can you just do

```suggestion
		const authorizeEndpoint = "https://api.intra.42.fr/oauth/authorize";
```

and make it a global variable (see providers like Apple)

File: src/providers/42.ts [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+
+export class FortyTwo implements OAuth2Provider {
+	private client: OAuth2Client;
+	private clientSecret: string;
+
+	constructor(
+		clientId: string,
+		clientSecret: string,
+		options?: {
+			redirectURI?: string;
```
Comment (pilcrowonpaper) [inline src/providers/42.ts]: Is `redirectURI` required or not (the docs you wrote seems to indicate that it is)?

**Review by pilcrowonpaper — CHANGES_REQUESTED** [review]

**Comment by awbx** [comment]
include the modifications as suggested by @pilcrowOnPaper earlier.

**Comment by awbx** [comment]
ready to merge

File: src/providers/42.ts (hunk 3 above)
Comment (awbx) [reply src/providers/42.ts]: i can't because it's not compatible with the `validateAuthorizationCode` method result type.

Comment (pilcrowonpaper) [reply src/providers/42.ts]: I'm not sure what you mean?

```ts
const tokens: FortyTwoTokens = {
			accessToken: result.access_token,
			refreshToken: result.refresh_token ?? null, 
			expiresIn: result.expires_in ?? null,
}
```

Comment (awbx) [reply src/providers/42.ts]: my apologies, I've made the change as you suggested

**Comment by awbx** [comment]
ready to merge

File: src/providers/42.ts [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+			expiresIn: result.expires_in ?? null,
+			tokenType: result.token_type ?? null
+		};
+		return tokens;
+	}
+}
+
+export interface FortyTwoTokens {
+	accessToken: string;
+	refreshToken: string | null;
```
Comment (pilcrowonpaper) [inline src/providers/42.ts]: Is `refresh_token` and `expires_in` even included in the response?

Comment (pilcrowonpaper) [reply src/providers/42.ts]: Looks like `refresh_token` isn't included so you can just omit that, while `expires_in` is - so please replace it with `accessTokenExpiresAt` field (see other providers)

Comment (awbx) [reply src/providers/42.ts]: done

File: src/providers/42.ts [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+		return tokens;
+	}
+}
+
+interface TokenResponseBody {
+	access_token: string;
+	token_type: string;
+	expires_in: number;
+	scope: string;
+	create_at: number;
```
Comment (pilcrowOnPaper) [inline src/providers/42.ts:50]: Final fix - should be `created_at`!

Comment (pilcrowonpaper) [inline src/providers/42.ts]:
```suggestion
	created_at: number;
```

**Review by pilcrowonpaper — APPROVED** [review]

**Review by pilcrowonpaper — APPROVED** [review]

**Comment by pilcrowonpaper** [comment]
Thanks!

**Pull request merged**